Avoid updates (and update notifications) from appearing in more gameplay cases

peppy opened the pull request:
Second attempt at https://github.com/ppy/osu/pull/29657 taking on feedback from https://github.com/ppy/osu/pull/29657#discussion_r1738406567.

- [x] RFC, haven't tested update flow yet.

Closes #29341.

peppy committed:
Adjust `ILocalUserPlayInfo` to expose whether gameplay is in a paused/break state

peppy committed:
Make states better defined

peppy committed:
Avoid updates and update notifications appearing in more gameplay cases

peppy commented on osu.Game/OsuGame.cs:
```diff
@@ -391,11 +386,11 @@ private void load()
-                BeatmapManager.PauseImports = p.NewValue;
-                SkinManager.PauseImports = p.NewValue;
-                ScoreManager.PauseImports = p.NewValue;
+                BeatmapManager.PauseImports = p.NewValue != LocalUserPlayingStates.NotPlaying;
+                SkinManager.PauseImports = p.NewValue != LocalUserPlayingStates.NotPlaying;
+                ScoreManager.PauseImports = p.NewValue != LocalUserPlayingStates.NotPlaying;
```
Note that this stops imports from running during break time and on the fail screen, which is an extra win.

peppy committed:
Fix local scores never importing due to new conditionals

peppy committed:
Attempt to fix android class

bdach commented on osu.Game/Screens/Play/LocalUserPlayingStates.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace osu.Game.Screens.Play
+{
+    public enum LocalUserPlayingStates
```
Bit weird for this enum name to be in plural, I'd normally expect `LocalUserPlayingState`...?

peppy replied:
I originally had it plural to avoid the classic namespace conflict with field (was called `PlayingState`). Can go with non-plural not, sure.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -94,6 +94,7 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlayb
         public IBindable<bool> LocalUserPlaying => localUserPlaying;
```
I'd argue this should no longer be public (or even exist).

peppy replied:
Yeah, it's gonna touch a lot of tests though (take a quick look and see if you thing it's worth it)

bdach replied:
As far as I can see the extent of it is mostly asserts, right? If that is accurate then I think it's worth getting rid just to lower the redundancy around here. I'll have to cross-check all of the existing conditional changes anyway to make sure nothing slipped by so that seems like a rather minor inconvenience from both authoring and reviewing angles.

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -391,11 +386,12 @@ private void load()
-                ScoreManager.PauseImports = p.NewValue;
+                BeatmapManager.PauseImports = p.NewValue != LocalUserPlayingStates.NotPlaying;
+                SkinManager.PauseImports = p.NewValue != LocalUserPlayingStates.NotPlaying;
+                // For scores, we need to allow imports during "Break" state else local user's scores will never be imported.
+                ScoreManager.PauseImports = p.NewValue == LocalUserPlayingStates.Playing;
```
This is weird, can the state not transition to `NotPlaying` on gameplay completion or something?

peppy replied:
Can try and make that happen sure.

bdach reviewed (commented):
just a cursory read of structure, haven't examined the logic in detail yet (or tested)

peppy committed:
Passing means `NotPlaying`

peppy committed:
Rename enum to non-plural now that it won't conflict

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -508,7 +508,7 @@ private void onBreakTimeChanged(ValueChangedEvent<bool> isBreakTime)
         {
-            bool inGameplay = !DrawableRuleset.HasReplayLoaded.Value;
+            bool inGameplay = !DrawableRuleset.HasReplayLoaded.Value && !GameplayState.HasPassed;
             bool inBreak = breakTracker.IsBreakTime.Value || DrawableRuleset.IsPaused.Value || GameplayState.HasFailed;
```
Hmm with failing being considered as "break" I'm not sure the `ScoreManager.PauseImports` change is entirely correct. I'd say _any_ sort of gameplay completion should transition the state to `NotPlaying`...?

peppy replied:
The reason I didn't originally add failing is that I thought we'd want things like imports to stay blocked while at the fail screen (else they may continue into the start of the next play). I think it's also quite crippling when imports run during `HighPerformanceSession`. But maybe this is an edge case we just allow to happen for now?

bdach replied:
> we'd want things like imports to stay blocked while at the fail screen

Except for the part where you can request the failed play to be imported at the fail screen which will presumably deadlock now.

peppy replied:
Yeah, probably need a test and just doing as you say for now.

peppy committed:
Add failing test of importing failed replay in `OsuGame` flow

We had a test covering this but it wasn't within `OsuGame` so didn't
have full import blocking coverage (see
https://github.com/ppy/osu/blob/cbbe2f9dc04abb052a0d762aa798eba094fdd80c/osu.Game.Tests/Visual/Gameplay/TestScenePlayerLocalScoreImport.cs#L89-L88).

peppy committed:
Change failed states to be considered as `NotPlaying`

bdach commented:
Not sure of the status or urgency of this, are you looking to still check off the todo item in the OP or should I consider doing so part of review?

peppy commented:
I'll test this now, forgot.

peppy commented:
In testing it looks to work as expected.

Tested using inserted `Task.Delay()` to make sure I had enough time to get to gameplay. If you test, it's as simple as running `dotnet run` on `osu-deploy` and "installing" the game (however that is required by your testing platform).

We have downgrades enabled so it will detect the last published version and start to download / update to it.

smoogipoo commented on osu.Android/GameplayScreenRotationLocker.cs:
```diff
@@ -6,28 +6,29 @@
 {
     public partial class GameplayScreenRotationLocker : Component
     {
-        private Bindable<bool> localUserPlaying = null!;
+        private IBindable<LocalUserPlayingStates> localUserPlaying = null!;
```
What happened here? The class doesn't exist?

bdach replied:
Probably something something missed in the rename because `osu.Desktop.slnf`.

```suggestion
        private Bindable<bool> localUserPlaying = null!;
        private IBindable<LocalUserPlayingState> localUserPlaying = null!;
```

probably fixes.

peppy committed:
Fix android build failure due to enum rename

peppy committed:
Remove unnecessary `runOutsideOfGameplay` call

peppy committed:
Add second call of `runOutsideGameplay` on update progress notification

At this point the update is already started in the background but I
guess we can still block the notification from interrupting the user.

smoogipoo commented:
From a structural perspective, I really really don't like this one.

We have way too many states flying around right now:
- [`OsuScreen.OverlayActivationMode`](https://github.com/ppy/osu/blob/cf66d40b008ba70ec3037d11713bbaa0bda04e91/osu.Game/Screens/OsuScreen.cs#L55)
- [`OsuScreen.Activity`](https://github.com/ppy/osu/blob/cf66d40b008ba70ec3037d11713bbaa0bda04e91/osu.Game/Screens/OsuScreen.cs#L73)
- High performance session.
- ... and now yet another bindable that is messaging very similar things in a slightly different manner.

Resulting in a 19 file diff that is pretty rough to review.

Not mentioned here: what is a "local user playing state"? why is it not a "gameplay state"? we already have a [`GameplayState`](https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/GameplayState.cs) why can't that be used? do we really need to wrap one bindable in an interface like this?

While I'm going to grit my teeth and continue powering through this one given that it's the second attempt, in the future I suggest we use `SessionStatics` for this.

smoogipoo committed:
Merge branch 'master' into updates-outside-of-gameplay-only-2

smoogipoo committed:
Remove unrelated test

smoogipoo committed:
Refactoring for correctness

smoogipoo committed:
Fix incorrect condition

smoogipoo committed:
Return true while in gameplay

A `false` value marks the user as being on the latest release, and
notifies them as such when clicking the button in settings. In reality,
we don't know whether this is the case yet - we're just deferring the
check.

Somewhat minor change because the chance of a user manually going into
settings and clicking the button is very small.

smoogipoo committed:
Remove unused parameter

smoogipoo commented:
Have made a few changes here, please check the last few commits @peppy

smoogipoo commented on osu.Game/Input/OsuUserInputManager.cs:
```diff
@@ -10,7 +10,7 @@ namespace osu.Game.Input
     public partial class OsuUserInputManager : UserInputManager
     {
-        protected override bool AllowRightClickFromLongTouch => PlayingState.Value == LocalUserPlayingState.NotPlaying;
+        protected override bool AllowRightClickFromLongTouch => PlayingState.Value != LocalUserPlayingState.Playing;
```
Isn't the rest of the game interface technically available at the pause/fail screen? If you have a keyboard you can access settings, chat, user overlay, beatmap overlay, etc?

smoogipoo replied:
Also, as far as I can tell this is a revert to `master`. That said, I haven't gone through the other cases like `GameplayScreenRotationLocker` which change logic from `master` (again, as far as I can tell).

We discussed yesterday what would it take to make it so that PR reviews didn't take many times longer than actually making changes - not making subtle changes like this would help.

peppy replied:
Fine to go with as you have it, will revisit in a future.
